# <font color='blue'> Lab-04 (Solution) <font>


### Problem 1: Data Preparation
    
(1.0) Download any spam dataset of your choice. 
    
(1.1) Use Pandas to load the dataset and display the first few rows of the dataset.
    
(1.2) Calculate the frequency of spam and ham messages in the dataset to assess the data balance.
    
<font color='brown'> Hint: You can download the data from here: https://archive.ics.uci.edu/dataset/228/sms+spam+collection<font>

In [2]:
#!pip install pandas --quiet

In [12]:
import pandas as pd

# (1.1) Load the dataset 
#It is a text file with tab-separated values,specify the separator as \t with sep='\t'.
path = "/Users/johannuchelwildt/Downloads/sms+spam+collection/SMSSpamCollection"

d_frame = pd.read_csv(path, sep='\t', header=None, names=['label', 'message'])

# Display the first 9 rows of the dataset
print(d_frame.head(9))

# (1.2) Count the number of spam and ham messages
class_balance= d_frame['label'].value_counts()
print("\n", class_balance)

  label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
5  spam  FreeMsg Hey there darling it's been 3 week's n...
6   ham  Even my brother is not like to speak with me. ...
7   ham  As per your request 'Melle Melle (Oru Minnamin...
8  spam  WINNER!! As a valued network customer you have...

 label
ham     4825
spam     747
Name: count, dtype: int64


In [13]:
d_frame.describe()

,label,message
count,5572,5572
unique,2,5169
top,ham,"Sorry, I'll call later"
freq,4825,30


### Problem 2: Data Pre-processing
    
Create a function 'data_preprocess(text)' in NLTK that takes the text data as input and returns the pre-processed word tokens as 'new_text'. Inside the function apply the following pre-processing techniques:  
    
(2.1) Convert input text into lowercase.
    
(2.2) Tokenize the text into word tokens.
    
(2.3) Remove stopwords.
    
(2.4) Apply stemming or lemmatization into word tokens.
    
Preview the data to ensure pre-processing is as expected.

In [14]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

# Download stopwords from NLTK
#nltk.download('punkt')
#nltk.download('stopwords')

def data_preprocess(text):
    # (2.1) lowercase the text
    text = text.lower()     
    # (2.2) Tokenize the text into words
    tokens = word_tokenize(text)
    new_tokens=[]
    # (2.3)  Remove stopwords
    ps = PorterStemmer()
    for w in tokens:
        if w not in stopwords.words('english'):
    # (2.4) Stem the words
            stem_tokens = ps.stem(w)   
            new_tokens.append(stem_tokens)  # Add the stemmed word to new_tokens
    return  new_tokens

# Apply preprocessing to each message
d_frame['message'] = d_frame['message'].apply(data_preprocess)

# Preview the data to ensure pre-processing is as expected
print(d_frame.head(10))


  label                                            message
0   ham  [go, jurong, point, ,, crazi, .., avail, bugi,...
1   ham             [ok, lar, ..., joke, wif, u, oni, ...]
2  spam  [free, entri, 2, wkli, comp, win, fa, cup, fin...
3   ham  [u, dun, say, earli, hor, ..., u, c, alreadi, ...
4   ham  [nah, n't, think, goe, usf, ,, live, around, t...
5  spam  [freemsg, hey, darl, 's, 3, week, 's, word, ba...
6   ham  [even, brother, like, speak, ., treat, like, a...
7   ham  [per, request, 'mell, mell, (, oru, minnaminun...
8  spam  [winner, !, !, valu, network, custom, select, ...
9  spam  [mobil, 11, month, ?, u, r, entitl, updat, lat...


### <font color='white'> Problem 3: Feature Extraction and data split: <font>
    
(3.1) Convert text data into numerical features by using the Bag of Words. Use the top 1000 words as your features.
    
 (3.2) Split the dataset into train (75%) and test (25%) sets.

In [6]:
#(3.1)

# Convert DataFrame rows into a list of tuples, each containing message and label
ml_list = [(m,l) for m,l in zip(d_frame['message'], d_frame['label'])]

# Create a list of all word tokens
w_list = []
for w in d_frame['message']:
    w_list.extend(w)

# Create a frequency distribution of all words
w_count = nltk.FreqDist(w_list)

# Use the top 1000 words as features
top_words = list(w_count.keys())[:1000]

# Define a Feature Extraction Function
def extract_features(doc):
    #convert the doc into a set of unique words
    w_set = set(doc)
    # initializes an empty dictionary
    f_dict = {}
    for w in top_words:
        # checks whether that word is present in the words set. 
        # The result is a Boolean value (True if the word is present, False otherwise)
        f_dict[w] = True if w in w_set else False  
    return f_dict

# Created Feature Sets
f_sets = [(extract_features(m),l) for (m,l) in ml_list]

# (3.2) Split 75% of the data for training, 25% for testing
size = int(len(f_sets)*0.75)
train_set, test_set = f_sets[:size], f_sets[size:]

In [7]:
ml_list

[(['go',
   'jurong',
   'point',
   ',',
   'crazi',
   '..',
   'avail',
   'bugi',
   'n',
   'great',
   'world',
   'la',
   'e',
   'buffet',
   '...',
   'cine',
   'got',
   'amor',
   'wat',
   '...'],
  'ham'),
 (['ok', 'lar', '...', 'joke', 'wif', 'u', 'oni', '...'], 'ham'),
 (['free',
   'entri',
   '2',
   'wkli',
   'comp',
   'win',
   'fa',
   'cup',
   'final',
   'tkt',
   '21st',
   'may',
   '2005.',
   'text',
   'fa',
   '87121',
   'receiv',
   'entri',
   'question',
   '(',
   'std',
   'txt',
   'rate',
   ')',
   '&',
   'c',
   "'s",
   'appli',
   '08452810075over18',
   "'s"],
  'spam'),
 (['u',
   'dun',
   'say',
   'earli',
   'hor',
   '...',
   'u',
   'c',
   'alreadi',
   'say',
   '...'],
  'ham'),
 (['nah', "n't", 'think', 'goe', 'usf', ',', 'live', 'around', 'though'],
  'ham'),
 (['freemsg',
   'hey',
   'darl',
   "'s",
   '3',
   'week',
   "'s",
   'word',
   'back',
   '!',
   "'d",
   'like',
   'fun',
   'still',
   '?',
   'tb',
   'ok',


In [ ]:
# another simpler solution that uses sklearn modules - making it a lil' simpler

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import pandas as pd

# Example: Assuming d_frame['message'] is tokenized like [['this', 'is', 'a', 'message'], ['another', 'message']]
vectorizer = TfidfVectorizer(max_features=1000, analyzer='word', preprocessor=lambda x: x, tokenizer=lambda x: x)
X = vectorizer.fit_transform(d_frame['message']).toarray()
y = d_frame['label']

# Split the dataset into train (75%) and test (25%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

print("Training set size:", len(X_train))
print("Test set size:", len(X_test))


Training set size: 4179
Test set size: 1393


/Users/johannuchelwildt/Library/Python/3.9/lib/python/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


--- 
###  Problem 4: Naive Bayes Classifier Training
    
(4.1) Train a Naive Bayes classifier using the features extracted in Problem 3. 
 
(4.2) Show most informative 20 features. 
    
(4.3) Predict the labels for the test set.

In [ ]:
from nltk import NaiveBayesClassifier # For creating and using a Naive Bayes classifier
from collections import defaultdict # For dictionaries with default values for missing keys

# (4.1) Train the Naive Bayes Classifier
nb = NaiveBayesClassifier.train(train_set)

#(4.2) Most informative features
nb.show_most_informative_features(20)

# (4.3) Get the list of actual class from the test set and predicted class from the classifier
predicted_label=[]
actual_label=[]
for f, l in test_set:
    predicted_label.append(nb.classify(f)) 
    actual_label.append(l)

Most Informative Features
                   award = True             spam : ham    =    185.2 : 1.0
                   nokia = True             spam : ham    =    168.2 : 1.0
                  servic = True             spam : ham    =    144.3 : 1.0
                     txt = True             spam : ham    =    109.4 : 1.0
                      16 = True             spam : ham    =     87.3 : 1.0
                    code = True             spam : ham    =     83.0 : 1.0
                deliveri = True             spam : ham    =     66.0 : 1.0
                 landlin = True             spam : ham    =     65.1 : 1.0
                      po = True             spam : ham    =     65.1 : 1.0
                    rate = True             spam : ham    =     60.0 : 1.0
                  camera = True             spam : ham    =     57.5 : 1.0
                   mobil = True             spam : ham    =     56.8 : 1.0
                  urgent = True             spam : ham    =     52.8 : 1.0

### Problem 5: Model Evaluation:
         
(5.1) Calculate the accuracy of the model.
     
(5.2) Calculate precision, recall, F1-score for both classes (spam and ham) and print them in table format.

In [11]:
import nltk.metrics
from nltk.metrics import precision, recall, f_measure, accuracy
from collections import defaultdict


# (5.1) Accuracy
a = round(accuracy(actual_label, predicted_label),2)
print(f"Accuracy of the Naive Bayes classifier is {a*100}%. \n")


#(5.2) Precision, Recall, F-Measure

# key -> label 
# refsets value -> set of indices representing the test instances that truly belong to that class.
# checksets: value -> set of indices of test instances that were predicted to belong to the class.
refsets = defaultdict(set)
checksets = defaultdict(set)

for i, (f, l) in enumerate(test_set):
    refsets[l].add(i)
    observed = nb.classify(f)
    checksets[observed].add(i)

precision_spam = round(precision(refsets['spam'], checksets['spam']),2)
recall_spam = round(recall(refsets['spam'], checksets['spam']),2)
f_measure_spam = round(f_measure(refsets['spam'], checksets['spam']),2)

# Precision, Recall, F-Measure for neg classes

precision_ham = round(precision(refsets['ham'], checksets['ham']),2)
recall_ham = round(recall(refsets['ham'], checksets['ham']),2)
f_measure_ham = round(f_measure(refsets['ham'], checksets['ham']),2)

#  Print in table format using Panda library <font>

row_labels = ['Spam', 'Ham']

table = {
    ' Precision(%)': [precision_spam*100, precision_ham*100],
    ' Recall(%)': [recall_spam*100, recall_ham*100],
    ' F1-Measure(%)': [f_measure_spam*100, f_measure_ham*100]
}
df = pd.DataFrame(table, index=row_labels)
print(df)

Accuracy of the Naive Bayes classifier is 98.0%. 

      Precision(%)  Recall(%)  F1-Measure(%)
Spam          98.0       87.0           92.0
Ham           98.0      100.0           99.0


Spam
* ⁠Precision = 0.98 → 98% of messages predicted as spam were actually spam. => TP/(TP+FP)
* ⁠Recall = 0.87 → 87% of actual spam messages were correctly classified. => TP / (TP+FN)
* ⁠F1-score = 0.92 → A balanced measure of precision and recall.

### <font color='white'> Bonus Question: (Extra) <font>
    
Try to develop the similar classification model and use TF-IDF method for feature extraction in scikit-learn. 

<font color='brown'> Hint: See the tutorial here https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html#   <font>